ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,415 @@ let order_f (_, vm_rec) =
+    pool_state
+
+(* The no breach plan pool state determines the no breach evacuation planning
+   for anti-affinity VMs.
+   It's a VMGrpMap which contains "no breach plan state" for each VM anti-
+   affinity group.
+   "no breach plan state" has 2 elements:
+   1. a Psq which contains "no_resident" hosts for that group. (A "no
+      resident" host for a group is a host which has no running VMs from that
+      group resident on it.)
+   2. an int which is the number of "resident" hosts for each group. (A
+      "resident" host for a group is a host which has at least one running VM
+      from that group resident on it.)
+   Let's say the anti-affinity VM to be planned belongs to anti-affinity group
+   A. If for group A, the number of "resident" hosts is already 2 or greater
+   than 2, then we don't need to plan the VM on any host, if not, we will need
+   to check the host with the minimal free memory from the "no resident" hosts
+   queue, which is just the minimal host returned from "Psq.min" on the "no
+   resident" hosts Psq of group A. *)
+let init_no_breach_plan_pool_state spread_evenly_plan_pool_state =
+  let module Q = AntiAffinityEvacPlanHostPsq in
+  spread_evenly_plan_pool_state
+  |> VMGrpMap.map (fun hs ->
+         let no_resident_hosts, resident_hosts =
+           Q.partition (fun _ (vm_cnt, _) -> vm_cnt = 0) hs
+         in
+         (no_resident_hosts, Q.size resident_hosts)
+     )
+
+(* Update "no_breach_plan_pool_state" after a VM from anti-affinity "group"
+   with memory size: "vm_size" is planned on the "host":
+   1. For the "group", the "host" is removed from its "no_resident" hosts
+      queue, and increase its "resident_hosts_cnt" by 1.
+   2. For other groups, updates the host's size by substracting "vm_size" if
+      the host is in that group's "no_resident" hosts queue. *)
+let update_no_breach_plan_pool_state vm_size group host pool_state =
+  let module Q = AntiAffinityEvacPlanHostPsq in
+  VMGrpMap.mapi
+    (fun grp (no_resident_hosts, resident_hosts_cnt) ->
+      match grp = group with
+      | true ->
+          (Q.remove host no_resident_hosts, succ resident_hosts_cnt)
+      | false ->
+          ( Q.update host
+              (Option.map (fun (vm_cnt, h_size) ->
+                   (vm_cnt, Int64.sub h_size vm_size)
+               )
+              )
+              no_resident_hosts
+          , resident_hosts_cnt
+          )
+    )
+    pool_state
+
+(* For an anti-affinity group, select host for a VM of memory size: vm_size
+   from hosts Psq: hosts_psq, returns the selected host for the VM and the
+   available hosts Psq for the remaining VMs to be planned in that anti-
+   affinity group. *)
+let rec select_host_for_anti_aff_evac_plan vm_size hosts_psq =
+  let module Q = AntiAffinityEvacPlanHostPsq in
+  match Q.pop hosts_psq with
+  | None ->
+      None
+  | Some ((host, (_, h_size)), rest_hosts) -> (
+    match vm_size <= h_size with
+    | true ->
+        Some (host, hosts_psq)
```

Comment:
> This relies on the fact that the list of VMs is sorted from small to large...

Yes. And the checking is only on the VM size. E.g. in previous PR, it would check the `can_boot_on`, then the current improvement would not work, as even if the current VM couldn't boot on a host, a bigger VM might do.

Reply:
Yes, this is only an optimization: if a host is smaller than the current VM, it doesn't need to be considered for the next VM.
Added some comments to make it clear.

Reply:
> I personally would have chosen different names for some of the values, but the comments make things clearer.
@robhoes , would you please point those names and a better way for them? Not only for this time but also will be better for latter PRs, thanks~